External/SPMmodified/xASL/xASL_im_FillNaNs.m

```diff
@@ -158,29 +183,67 @@
 
 %% ========================================================================================
 %% ========================================================================================
-function [IM] = xASL_im_ExtrapolateLinearlyOverNaNs(IM)
+%% METHOD 3
+function [IM] = xASL_im_ExtrapolateLinearlyOverNaNs(IM, nEdgeVoxels)
 %xASL_im_ExtrapolateLinearlyOverNaNs
 % The input of this function is 3D (iterations of other dimensions are done
 % in the main function above)
+%
+% IM          - input image
+% nEdgeVoxels - number of voxels at the edges of the non-nan values that we check for weird interpolation values, 
+% to be set to NaN and ignored in the extrapolation
 
-% Zeros are also converted to NaNs - because this is sometimes unclear in the flow fields.
-IM(IM==0) = NaN;
+%% 0. Admin
+% Also catch Infs or imaginary numbers or whatever. Zeroes are already catched above in the main function.
 IM(~isfinite(IM)) = NaN;
 
+if nargin<2 || isempty(nEdgeVoxels)
+    nEdgeVoxels = 0;
+    % We can try this later, but it seems to work less well than doing the same based on the distance transform
+end
+
+
+%% 0.5 Fix interpolation edges
+% The first layer(s) next to the NaNs are sometimes interpolated as well
+% So here we try to estimate which voxels those are, by looking at the distance from the NaNs
+% together with the transformation difference from a smoothed transformation field
+% Note that this could also potentially fix extreme transformations
+
 % Get distance transform
-distMap = xASL_im_DistanceTransform(~isfinite(IM));
+distMap = xASL_im_DistanceTransform(isnan(IM));
+
+% Smooth the image outside of the NaNs
+IMsmooth = xASL_im_ndnanfilter(IM, 'gauss', [8 8 8], 1);
+% Get an absolute difference image between the original and the smoothed transformation field
```

Comment:
I'm a bit confused what is happening here and why. I thought we said we'll replace NaNs by identity.
line 213-233 are very complex and I'm really not sure if we need anything like this.

Reply:
Yes - I used the "identity" parallel, but indeed meant the "extrapolated identity".

OK. I will check the code and see if we can get rid of those outliers before having them, or otherwise keep your solution and clean it.